Add validation for Tool annotation names

academey opened the pull request:
## Summary

This PR adds validation for tool names in the `@Tool` annotation to prevent runtime failures when Large Language Models (LLMs) encounter invalid characters.

## Problem

As reported in #3832, tool names containing spaces or special characters can cause LLMs like GPT 4.1-mini to fail during execution. Currently, there is no validation to prevent these invalid names from being used.

## Solution

- Added regex pattern validation in `ToolUtils.getToolName()` method
- Tool names can now only contain:
  - Alphanumeric characters (a-z, A-Z, 0-9)
  - Underscores (_)
  - Hyphens (-)
  - Dots (.)
- Throws `IllegalArgumentException` with a clear error message when invalid characters are detected
- Updated `@Tool` annotation documentation to clarify naming constraints

## Changes

1. **ToolUtils.java**: Added validation logic with regex pattern `^[a-zA-Z0-9_\\.-]+$`
2. **Tool.java**: Enhanced JavaDoc with naming constraints and examples
3. **ToolUtilsTests.java**: Added comprehensive unit tests covering:
   - Valid tool names
   - Invalid names with spaces
   - Invalid names with special characters
   - Invalid names with parentheses
   - Edge cases (empty names, methods without annotations)

## Testing

- All new tests pass successfully
- Existing tests continue to work
- No breaking changes for valid tool names

## Example

```java
// Valid tool names
@Tool(name = "get_weather")
@Tool(name = "search-docs")
@Tool(name = "tool.v1")

// Invalid tool names (will throw IllegalArgumentException)
@Tool(name = "get weather")  // contains space
@Tool(name = "tool()")       // contains parentheses
@Tool(name = "tool@name!")   // contains special characters
```

Fixes #3832

Signed-off-by: Hyunjoon Park <academey@gmail.com>

rodrigomalara committed:
[GH-3723] Vertex AI Gemini logprobs support

Signed-off-by: Rodrigo Malara <rodrigomalara@gmail.com>

sunyuhan1998 committed:
refactor: GH-3620 Add `BedrockChatOptions` to Bedrock

Signed-off-by: Sun Yuhan <sunyuhan1998@users.noreply.github.com>

tzolov committed:
feat: add streamable-HTTP transport support for MCP client

- Split MCP client auto-configuration into transport-specific modules to support new streamable-HTTP transport
  - spring-ai-autoconfigure-mcp-client-common (shared functionality)
  - spring-ai-autoconfigure-mcp-client-httpclient (HttpClient-based transports)
  - spring-ai-autoconfigure-mcp-client-webflux (WebFlux-based transports)
- Update starters to use appropriate transport-specific modules
- Add streamable-HTTP configuration properties (spring.ai.mcp.client.streamable-http)
- Update documentation with streamable-HTTP transport examples
- Update BOM to reflect new modular structure

Signed-off-by: Christian Tzolov <christian.tzolov@broadcom.com>

markpollack committed:
Extract VectorStoreRetriever interface from VectorStore (#3827)

- Created new VectorStoreRetriever functional interface for read-only operations
- Modified VectorStore to extend VectorStoreRetriever
- Removed duplicate similaritySearch methods from VectorStore
- Updated API documentation to reflect the interface separation
- Added package-info.java documentation for clarity

This change follows the principle of least privilege by providing a read-only
interface for retrieval operations while maintaining the full functionality
in VectorStore.

Fixes #1290

Signed-off-by: Mark Pollack <mark.pollack@broadcom.com>

finyuq committed:
fix method tool call throws same method name (#2774)

Auto-cherry-pick to 1.0.x
Fixes #2771
Signed-off-by: finyu <18664388610@msn.cn>

ilayaperumalg committed:
Refactor MethodToolCallbackProviderTests

 - Move the existing buildToolsWithBridgeMethodReturnOnlyUserDeclaredMethods into MethodToolCallbackProviderTests
 - Delete MethodToolCallbackProviderTest

Signed-off-by: Ilayaperumal Gopinathan <ilayaperumal.gopinathan@broadcom.com>

Skillkiller committed:
Fix PropmptTemplate typo in structured-output-api.jpg (#3304)

Auto-cherry-pick to 1.0.x
Fixes #3304

Signed-off-by: Skillkiller <seblsebastian@aol.de>

academey committed:
Add validation for Tool annotation names

- Add regex pattern validation for tool names in ToolUtils
- Tool names can only contain alphanumeric characters, underscores, hyphens, and dots
- Add comprehensive unit tests for tool name validation
- Update @Tool annotation documentation with naming constraints
- Throw IllegalArgumentException for invalid tool names with clear error message

This prevents runtime failures when LLMs (like GPT 4.1-mini) encounter tool names
with spaces or special characters that they cannot process.

Fixes #3832

Signed-off-by: Hyunjoon Park <academey@gmail.com>

sunyuhan1998 commented on spring-ai-model/src/main/java/org/springframework/ai/tool/support/ToolUtils.java:
```diff
@@ -38,16 +39,27 @@
+	 * Regular expression pattern for valid tool names. Tool names can only contain
+	 * alphanumeric characters, underscores, hyphens, and dots.
+	 */
+	private static final Pattern TOOL_NAME_PATTERN = Pattern.compile("^[a-zA-Z0-9_\\.-]+$");
```
Hi, thank you for paying attention to this issue and trying to resolve it. I admit that such a restriction is a standardized practice, but considering that `Tool` is a generic annotation — it doesn't serve solely openAI or a specific model — limiting tool names to [a-z][A-Z] might be unfair to models that support a wide variety of tool names. This is just my personal opinion; I hope it doesn't bother you. Thank you again.

academey replied:
Thank you for the valuable feedback! 
I agree that enforcing a strict validation pattern could be too restrictive for models that support broader character sets. Let me propose a more flexible approach:

Warning-based approach
 ```java
  // Log a warning instead of throwing an exception
  if (!RECOMMENDED_PATTERN.matcher(toolName).matches()) {
      logger.warn("Tool name '{}' may not be compatible with some LLMs (e.g., OpenAI). " +
                  "Consider using only alphanumeric characters, underscores, hyphens, and dots.",
                  toolName);
  }
```
What do you think about these approach? I'm happy to update the PR

sunyuhan1998 replied:
> What do you think about these approach? I'm happy to update the PR

I think the approach you mentioned is a good idea, at least for now, until we have a more refined process and conduct dedicated tool name compatibility checks for each model. It can at least provide users with some troubleshooting ideas in case they really encounter issues with tool name incompatibility.

academey replied:
I've updated the implementation based on the feedback. The validation now logs a warning instead of throwing an exception, making it more flexible while still providing guidance to users.

This approach maintains backward compatibility and allows the Tool annotation to remain generic and LLM-agnostic, while still helping users avoid potential compatibility issues.

Thanks for your feedback @sunyuhan1998

academey committed:
Change Tool name validation from exception to warning

Based on feedback, the validation has been changed from throwing an exception
to logging a warning. This approach is more flexible as the Tool annotation
is generic and not specific to any particular LLM.

Changes:
- Modified ToolUtils to log warnings instead of throwing exceptions
- Updated Tool annotation JavaDoc to recommend naming conventions
- Adjusted tests to verify tool names are returned (no exceptions thrown)
- Added test for unicode characters to support non-English contexts

The warning message guides users toward compatible naming while allowing
flexibility for different LLMs and use cases.

Fixes spring-projects#3832

Signed-off-by: Hyunjoon Park <academey@gmail.com>
